extension.ts

```diff
@@ -0,0 +1,303 @@
+const SOMOD = ".somod";
+const FUNCTION_LAYERS = "functionLayers";
+const ROUTE_FILE_NAME = "routes.js";
+
+export const prebuild: Hook = async (context: IContext) => {
+  const rootModule = context.moduleHandler.getModule(
+    context.moduleHandler.roodModuleName
+  );
+  const routes = await getRouteFiles(rootModule);
+
+  for (const route of routes) {
+    /**
+     * TODO: check the schema of yaml file
+     */
+
+    const routePath = path.join(
+      rootModule.module.packageLocation,
+      SERVERLESS,
+      FUNCTIONS,
+      route + HTTP_YAML
+    );
+
+    const doc = await readYamlFileStore(routePath);
+    const violations: Violation[] = await validate(schema, doc);
+    if (violations.length > 0) {
+      console.log("error - " + violations[0].message);
+      throw violations[0].message;
+    }
+  }
+};
+
+export const build: Hook = async (context: IContext) => {
+  const rootModule = context.moduleHandler.getModule(
+    context.moduleHandler.roodModuleName
+  );
+  const routes = await getRouteFiles(rootModule);
+
+  /**
+   * TODO: convert the yaml file to json file and put it inside build folder
+   */
+
+  for (const route of routes) {
+    const routePath = path.join(
+      rootModule.module.packageLocation,
+      SERVERLESS,
+      FUNCTIONS,
+      route + HTTP_YAML
+    );
+
+    const routeJsonPath = path.join(
+      rootModule.module.packageLocation,
+      BUILD,
+      SERVERLESS,
+      FUNCTIONS,
+      route + HTTP_JSON
+    );
+
+    const doc = await readYamlFileStore(routePath);
+
+    /**
+     * TODO: merge prebuild and build
+     */
+    await writeFile(routeJsonPath, JSON.stringify(doc));
+  }
+};
+
+export const prepare: Hook = async (context: IContext) => {
+  console.log("prepare called");
+  const root = context.moduleHandler.getModule(
+    context.moduleHandler.roodModuleName
+  );
+
+  /**
+   * TODO:parse template.yaml and find functions
+   * then find module names and go to modules and find getconfigs
+   * then generate schema compiled files and save inside .somod
+   * create the layer in template.yaml and give contentUri path of previous step.
+   */
+  const templateFile = path.join(root.module.packageLocation, TEMPLATE_YAML);
+  const template: JSONObjectType = await readYamlFileStore(templateFile);
+
+  const functions = parseFunctions(template);
+
+  await Promise.all(
+    functions.map(async (func: JSONObjectType) => {
+      const props = func[PROPERTIES] as JSONObjectType;
+      let codeUri = props[CODE_URI] as string;
+      codeUri = codeUri.replace(
+        path.join(SOMOD, SERVERLESS, FUNCTIONS) + "/",
+        ""
+      );
+
+      // get module name
+      /**
+       * TODO: recheck this syntax
+       */
+      const ps = codeUri.split("/");
+      ps.pop();
+      const moduleName = ps.join("/");
+
+      //get module path from name
+      const module = context.moduleHandler.getModule(moduleName);
+      const __routeFileNames = await getRouteFiles(module);
```

Comment:
Update `getRouteFiles` to read routes from source and build

Reply:
done